File: frontend/src/objects/shared/access/AccessUserGroupSlug.ts
/** 访问策略中的用户组 slug；独立于用户组领域类型，避免跨边界误用。 */
export type AccessUserGroupSlug = string & { readonly __brand: 'AccessUserGroupSlug' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

const userGroupSlugPattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/

/** 将访问策略用户组 slug 还原为 URL/API 可传输字符串；无副作用。 */
export function accessUserGroupSlugValue(slug: AccessUserGroupSlug): string {
  return slug
}

/** 解析并校验访问策略用户组 slug；返回结构化失败原因，不抛异常。 */
export function parseAccessUserGroupSlug(rawSlug: string): ParseResult<AccessUserGroupSlug> {
  const normalized = rawSlug.trim()

  if (!normalized) {
    return { ok: false, error: 'User group slug is required.' }
  }

  if (normalized.length < 3 || normalized.length > 64) {
    return { ok: false, error: 'User group slug must be between 3 and 64 characters.' }
  }

  if (!userGroupSlugPattern.test(normalized)) {
    return { ok: false, error: 'User group slug may contain only lowercase letters, numbers, and hyphens.' }
  }

  /** 注意：这里的 as 只在完成 slug 格式校验后施加品牌类型，不改变运行时值。 */
  return { ok: true, value: normalized as AccessUserGroupSlug }
}

File: backend/src/main/scala/shared/objects/access/AccessUserGroupSlug.scala
package shared.objects.access

/** 访问控制中引用的用户组 slug，独立于用户组领域对象以降低共享层依赖。 */
final case class AccessUserGroupSlug(value: String)

/** 提供授权用户组 slug 的输入校验，确保共享访问策略不接收非法路径片段。 */
object AccessUserGroupSlug:
  private val slugPattern = "^[a-z0-9]+(?:-[a-z0-9]+)*$".r

  /** 解析用户组 slug，要求小写字母、数字和中划线组成，返回业务校验错误。 */
  def parse(raw: String): Either[String, AccessUserGroupSlug] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("User group slug is required.")
    else if normalized.length < 3 || normalized.length > 64 then Left("User group slug must be between 3 and 64 characters.")
    else if !slugPattern.matches(normalized) then Left("User group slug may contain only lowercase letters, numbers, and hyphens.")
    else Right(AccessUserGroupSlug(normalized))
